File: server/src/pty-manager.ts
/**
 * PtyManager — manages a Map of PTY instances (nvim editors and $SHELL terminals).
 *
 * Each instance is identified by a PtyInstanceId (UUID v4 generated client-side).
 * Editor PTYs spawn nvim with a --listen socket; terminal PTYs spawn $SHELL.
 * The first editor PTY is tracked as "primary" for buffer-watch and /api/nvim-open.
 */

import { spawn, IPty } from 'node-pty';
import { MineoCfg, loadConfig } from './config';
import * as path from 'path';
import * as fs from 'fs';

type PtyInstanceId = string;
type PaneRole = 'neovim' | 'terminal' | 'monaco';

const CONFIG_PATH = process.env.MINEO_CONFIG || path.resolve(__dirname, '../../../config.json');
let cfg = loadConfig(CONFIG_PATH);

// Bundled nvim config lives at <repo-root>/nvim-config/
// __dirname at runtime: <root>/server/dist/ → two levels up = <root>/
const BUNDLED_CONFIG_DIR = path.resolve(__dirname, '../../nvim-config');

/**
 * Build the extra env vars to apply based on the configured nvim config mode.
 *  - 'system':  no override — nvim reads ~/.config/nvim as usual
 *  - 'bundled': point XDG_CONFIG_HOME at the in-app nvim-config directory
 *  - 'custom':  point XDG_CONFIG_HOME at the user-supplied configDir
 */
function nvimConfigEnv(): Record<string, string> {
  const mode = cfg.nvim.configMode ?? 'system';
  if (mode === 'bundled') {
    return { XDG_CONFIG_HOME: BUNDLED_CONFIG_DIR };
  }
  if (mode === 'custom' && cfg.nvim.configDir) {
    return { XDG_CONFIG_HOME: cfg.nvim.configDir };
  }
  return {};
}

/**
 * ISO 8613-6 colon-separated RGB → legacy semicolon-separated.
 * xterm.js only understands the semicolon form.
 */
const COLON_RGB_RE = /\x1b\[(\d+):2:(\d+):(\d+):(\d+)m/g;

type DataListener = (data: string) => void;

interface PtyInstance {
    pty: IPty;
    role: PaneRole;
    socketPath: string | undefined; // only for editor PTYs
    listeners: Set<DataListener>;
    disposed: boolean;
}

export class PtyManager {
    private instances = new Map<PtyInstanceId, PtyInstance>();
    private primaryId: PtyInstanceId | undefined;

    /**
     * Spawn a new PTY instance.
     * - role='neovim': spawns nvim with --listen socket
     * - role='terminal': spawns $SHELL (bash/zsh)
     */
    spawn(id: PtyInstanceId, role: PaneRole, cols: number, rows: number, cwd?: string): void {
        if (this.instances.has(id)) {
            throw new Error(`PTY instance ${id} already exists`);
        }

        // Use provided cwd if it's a valid absolute path that exists, else fall back to cfg.workspace
        const resolvedCwd = (cwd && cwd.startsWith('/') && fs.existsSync(cwd)) ? cwd : cfg.workspace;

        let pty: IPty;
        let socketPath: string | undefined;

        if (role === 'neovim') {
            socketPath = `/tmp/mineo-nvim-${id}.sock`;
            // Remove stale socket file before spawning
            try { fs.unlinkSync(socketPath); } catch { /* doesn't exist — fine */ }

            pty = spawn(cfg.nvim.bin, ['--listen', socketPath, '-c', 'set mouse=a'], {
                name: 'xterm-256color',
                cols,
                rows,
                cwd: resolvedCwd,
                env: {
                    ...process.env,
                    TERM: 'xterm-256color',
                    COLORTERM: 'truecolor',
                    ...nvimConfigEnv(),
                } as Record<string, string>,
            });

            // Track first editor as primary
            if (!this.primaryId) {
                this.primaryId = id;
            }
        } else {
            // Terminal — spawn user's shell
            const shell = process.env.SHELL || '/bin/bash';
            pty = spawn(shell, [], {
                name: 'xterm-256color',
                cols,
                rows,
                cwd: resolvedCwd,
                env: {
                    ...process.env,
                    TERM: 'xterm-256color',
                    COLORTERM: 'truecolor',
                } as Record<string, string>,
            });
        }

        const listeners: Set<DataListener> = new Set();
        const instance: PtyInstance = { pty, role, socketPath, listeners, disposed: false };

        // PTY stdout → listeners (with RGB translation for editor PTYs)
        pty.onData((data: any) => {
            if (instance.disposed) return;
            const str: string = typeof data === 'string' ? data : (data as Buffer).toString('utf8');
            const translated = role === 'neovim'
                ? str.replace(COLON_RGB_RE, '\x1b[$1;2;$2;$3;$4m')
                : str;
            for (const cb of instance.listeners) {
                cb(translated);
            }
        });

        // Handle PTY exit
        pty.onExit(() => {
            this.cleanup(id);
        });

        this.instances.set(id, instance);
    }

    /** Write data to a PTY's stdin. */
    write(id: PtyInstanceId, data: string | Buffer): void {
        const inst = this.instances.get(id);
        if (inst && !inst.disposed) {
            inst.pty.write(data as any);
        }
    }

    /** Resize a PTY. */
    resize(id: PtyInstanceId, cols: number, rows: number): void {
        const inst = this.instances.get(id);
        if (inst && !inst.disposed && cols > 0 && rows > 0) {
            inst.pty.resize(cols, rows);
        }
    }

    /**
     * Subscribe to data events from a PTY.
     * Returns an unsubscribe function (for ws-pty.ts compatibility).
     */
    onData(id: PtyInstanceId, cb: (data: string) => void): () => void {
        const inst = this.instances.get(id);
        if (!inst) return () => { /* no-op */ };
        inst.listeners.add(cb);
        return () => { inst.listeners.delete(cb); };
    }

    /** Kill a PTY instance. */
    kill(id: PtyInstanceId): void {
        const inst = this.instances.get(id);
        if (!inst) return;
        inst.disposed = true;
        try { inst.pty.kill(); } catch { /* already dead */ }
        this.cleanup(id);
    }

    /** Check if an instance exists. */
    has(id: PtyInstanceId): boolean {
        return this.instances.has(id);
    }

    /** Get the role of an instance. */
    getRole(id: PtyInstanceId): PaneRole | undefined {
        return this.instances.get(id)?.role;
    }

    /** Get the socket path for the primary (first editor) PTY. */
    getPrimarySocketPath(): string | undefined {
        if (!this.primaryId) return undefined;
        return this.instances.get(this.primaryId)?.socketPath;
    }

    /** Get the primary instance ID. */
    getPrimaryId(): PtyInstanceId | undefined {
        return this.primaryId;
    }

    /** Get the nvim binary path (for buffer-watch execFile). */
    getNvimBin(): string {
        return cfg.nvim.bin;
    }

    /** Get current nvim config info for the settings API. */
    getNvimConfigInfo(): { bin: string; configMode: string; configDir: string; bundledConfigDir: string } {
        return {
            bin: cfg.nvim.bin,
            configMode: cfg.nvim.configMode ?? 'system',
            configDir: cfg.nvim.configDir ?? '',
            bundledConfigDir: BUNDLED_CONFIG_DIR,
        };
    }

    /**
     * Reload config — accepts an already-loaded MineoCfg snapshot,
     * or re-reads from disk if called with no arguments.
     */
    reloadConfig(freshCfg?: MineoCfg): void {
        const next = freshCfg ?? loadConfig(CONFIG_PATH);
        cfg.nvim.bin = next.nvim.bin;
        cfg.nvim.configMode = next.nvim.configMode;
        cfg.nvim.configDir = next.nvim.configDir;
    }

    /** Get the socket path for an instance. */
    getSocketPath(id: PtyInstanceId): string | undefined {
        return this.instances.get(id)?.socketPath;
    }

    /** Dispose all PTY instances — called on server shutdown. */
    disposeAll(): void {
        for (const [id] of this.instances) {
            this.kill(id);
        }
    }

    private cleanup(id: PtyInstanceId): void {
        const inst = this.instances.get(id);
        if (!inst) return;
        inst.disposed = true;
        inst.listeners.clear();
        // Clean up socket file for editor PTYs
        if (inst.socketPath) {
            try { fs.unlinkSync(inst.socketPath); } catch { /* fine */ }
        }
        this.instances.delete(id);
        // If primary was killed, promote next editor
        if (this.primaryId === id) {
            this.primaryId = undefined;
            for (const [nextId, nextInst] of this.instances) {
                if (nextInst.role === 'neovim' && !nextInst.disposed) {
                    this.primaryId = nextId;
                    break;
                }
            }
        }
    }
}

export const ptyManager = new PtyManager();

File: server/pty_manager.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"sync"

	"github.com/creack/pty"
)

// PaneRole describes the kind of PTY instance.
type PaneRole string

const (
	RoleNeovim   PaneRole = "neovim"
	RoleTerminal PaneRole = "terminal"
)

// DataListener receives PTY output as raw bytes.
type DataListener func(data []byte)

// PtyInstance represents a single running PTY process.
type PtyInstance struct {
	ptmx       *os.File
	cmd        *exec.Cmd
	role       PaneRole
	socketPath string // only for neovim PTYs
	listeners  map[string]DataListener
	listenerMu sync.Mutex
	disposed   bool
	nextListID int
}

// PtyManager manages a collection of PTY instances.
type PtyManager struct {
	mu        sync.RWMutex
	instances map[string]*PtyInstance
	primaryID string
	cfg       *MineoCfg
	appDir    string // directory of the running binary (for bundled config)
}

// NewPtyManager creates a new PTY manager.
func NewPtyManager(cfg *MineoCfg, appDir string) *PtyManager {
	return &PtyManager{
		instances: make(map[string]*PtyInstance),
		cfg:       cfg,
		appDir:    appDir,
	}
}

// colonCSIRe matches any CSI sequence that contains colon-separated subparams.
// xterm.js only understands semicolons as param separators; colons cause parse errors.
// Pattern: ESC [ <params-with-colons> <final-byte>
// We replace every colon inside the param area with a semicolon.
var colonCSIRe = regexp.MustCompile(`\x1b\[([0-9:;]*)([A-Za-z])`)

// nvimConfigEnv returns extra env vars based on the configured nvim config mode.
func (pm *PtyManager) nvimConfigEnv() []string {
	pm.cfg.mu.RLock()
	mode := pm.cfg.Nvim.ConfigMode
	configDir := pm.cfg.Nvim.ConfigDir
	pm.cfg.mu.RUnlock()

	bundledConfigDir := filepath.Join(pm.appDir, "nvim-config")

	switch mode {
	case NvimConfigBundled:
		return []string{"XDG_CONFIG_HOME=" + bundledConfigDir}
	case NvimConfigCustom:
		if configDir != "" {
			return []string{"XDG_CONFIG_HOME=" + configDir}
		}
	}
	// system mode: no XDG_CONFIG_HOME override — nvim uses ~/.config/nvim
	return nil
}

// filterEnv removes a key from an env slice (used to strip inherited XDG_CONFIG_HOME).
func filterEnv(env []string, key string) []string {
	prefix := key + "="
	out := env[:0]
	for _, e := range env {
		if !strings.HasPrefix(e, prefix) {
			out = append(out, e)
		}
	}
	return out
}

// Spawn creates a new PTY instance.
func (pm *PtyManager) Spawn(id string, role PaneRole, cols, rows uint16, cwd string) error {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	if _, exists := pm.instances[id]; exists {
		return fmt.Errorf("PTY instance %s already exists", id)
	}

	// Resolve cwd: use provided if valid absolute path that exists, else workspace
	resolvedCwd := pm.cfg.Workspace
	if cwd != "" && filepath.IsAbs(cwd) {
		if info, err := os.Stat(cwd); err == nil && info.IsDir() {
			resolvedCwd = cwd
		}
	}

	// Base environment — ensure HOME is always set
	env := os.Environ()
	env = append(env, "TERM=xterm-256color", "COLORTERM=truecolor")
	if os.Getenv("HOME") == "" {
		env = append(env, "HOME="+homeDir())
	}

	var cmd *exec.Cmd
	var socketPath string

	pm.cfg.mu.RLock()
	nvimBin := pm.cfg.Nvim.Bin
	pm.cfg.mu.RUnlock()

	if role == RoleNeovim {
		socketPath = fmt.Sprintf("/tmp/mineo-nvim-%s.sock", id)
		// Remove stale socket
		os.Remove(socketPath)

		cmd = exec.Command(nvimBin, "--listen", socketPath, "-c", "set mouse=a")
		cmd.Dir = resolvedCwd

		// Apply nvim config env — strip any inherited XDG_CONFIG_HOME first
		// so system mode uses ~/.config/nvim cleanly
		env = filterEnv(env, "XDG_CONFIG_HOME")
		extraEnv := pm.nvimConfigEnv()
		env = append(env, extraEnv...)
		cmd.Env = env
	} else {
		shell := os.Getenv("SHELL")
		if shell == "" {
			shell = "/bin/bash"
		}
		cmd = exec.Command(shell)
		cmd.Dir = resolvedCwd
		cmd.Env = env
	}

	// Start the PTY
	winSize := &pty.Winsize{Rows: rows, Cols: cols}
	ptmx, err := pty.StartWithSize(cmd, winSize)
	if err != nil {
		return fmt.Errorf("failed to start PTY: %w", err)
	}

	inst := &PtyInstance{
		ptmx:       ptmx,
		cmd:        cmd,
		role:       role,
		socketPath: socketPath,
		listeners:  make(map[string]DataListener),
		disposed:   false,
	}

	// Track first editor as primary
	if role == RoleNeovim && pm.primaryID == "" {
		pm.primaryID = id
	}

	pm.instances[id] = inst

	// Read goroutine: read PTY output, translate colors for neovim, dispatch to listeners
	go func() {
		buf := make([]byte, 32*1024)
		for {
			n, err := ptmx.Read(buf)
			if n > 0 {
				data := make([]byte, n)
				copy(data, buf[:n])

				// Translate colon subparams to semicolons for neovim output.
				// Neovim emits ISO 8613-6 colon-separated SGR (e.g. \x1b[38:2:R:G:Bm)
				// which xterm.js cannot parse — it only understands semicolons.
				if role == RoleNeovim {
					data = colonCSIRe.ReplaceAllFunc(data, func(m []byte) []byte {
						// Only replace colons in the param region (before the final letter)
						last := len(m) - 1
						params := make([]byte, last)
						copy(params, m[1:last]) // strip leading ESC[
						for i, b := range params {
							if b == ':' {
								params[i] = ';'
							}
						}
						result := make([]byte, len(m))
						result[0] = 0x1b
						copy(result[1:last], params)
						result[last] = m[last]
						return result
					})
				}

				inst.listenerMu.Lock()
				for _, cb := range inst.listeners {
					cb(data)
				}
				inst.listenerMu.Unlock()
			}
			if err != nil {
				break
			}
		}
		// Process exited — clean up
		pm.cleanup(id)
	}()

	return nil
}

// Write sends data to a PTY's stdin.
func (pm *PtyManager) Write(id string, data []byte) {
	pm.mu.RLock()
	inst, ok := pm.instances[id]
	pm.mu.RUnlock()
	if ok && !inst.disposed {
		inst.ptmx.Write(data)
	}
}

// Resize changes a PTY's window size.
func (pm *PtyManager) Resize(id string, cols, rows uint16) {
	pm.mu.RLock()
	inst, ok := pm.instances[id]
	pm.mu.RUnlock()
	if ok && !inst.disposed && cols > 0 && rows > 0 {
		pty.Setsize(inst.ptmx, &pty.Winsize{Rows: rows, Cols: cols})
	}
}

// OnData subscribes to data events from a PTY.
// Returns an unsubscribe function.
func (pm *PtyManager) OnData(id string, cb DataListener) func() {
	pm.mu.RLock()
	inst, ok := pm.instances[id]
	pm.mu.RUnlock()
	if !ok {
		return func() {}
	}

	inst.listenerMu.Lock()
	listID := fmt.Sprintf("l%d", inst.nextListID)
	inst.nextListID++
	inst.listeners[listID] = cb
	inst.listenerMu.Unlock()

	return func() {
		inst.listenerMu.Lock()
		delete(inst.listeners, listID)
		inst.listenerMu.Unlock()
	}
}

// Kill terminates a PTY instance.
func (pm *PtyManager) Kill(id string) {
	pm.mu.Lock()
	inst, ok := pm.instances[id]
	if !ok {
		pm.mu.Unlock()
		return
	}
	inst.disposed = true
	pm.mu.Unlock()

	inst.ptmx.Close()
	if inst.cmd.Process != nil {
		inst.cmd.Process.Kill()
	}
	pm.cleanup(id)
}

// Has checks if an instance exists.
func (pm *PtyManager) Has(id string) bool {
	pm.mu.RLock()
	defer pm.mu.RUnlock()
	_, ok := pm.instances[id]
	return ok
}

// GetRole returns the role of an instance.
func (pm *PtyManager) GetRole(id string) (PaneRole, bool) {
	pm.mu.RLock()
	defer pm.mu.RUnlock()
	inst, ok := pm.instances[id]
	if !ok {
		return "", false
	}
	return inst.role, true
}

// GetPrimarySocketPath returns the socket path of the primary (first editor) PTY.
func (pm *PtyManager) GetPrimarySocketPath() string {
	pm.mu.RLock()
	defer pm.mu.RUnlock()
	if pm.primaryID == "" {
		return ""
	}
	inst, ok := pm.instances[pm.primaryID]
	if !ok {
		return ""
	}
	return inst.socketPath
}

// GetPrimaryID returns the primary instance ID.
func (pm *PtyManager) GetPrimaryID() string {
	pm.mu.RLock()
	defer pm.mu.RUnlock()
	return pm.primaryID
}

// GetNvimBin returns the configured nvim binary path.
func (pm *PtyManager) GetNvimBin() string {
	pm.cfg.mu.RLock()
	defer pm.cfg.mu.RUnlock()
	return pm.cfg.Nvim.Bin
}

// NvimConfigInfo holds nvim configuration details for the API.
type NvimConfigInfo struct {
	Bin              string `json:"bin"`
	ConfigMode       string `json:"configMode"`
	ConfigDir        string `json:"configDir"`
	BundledConfigDir string `json:"bundledConfigDir"`
}

// GetNvimConfigInfo returns current nvim config info for the settings API.
func (pm *PtyManager) GetNvimConfigInfo() NvimConfigInfo {
	pm.cfg.mu.RLock()
	defer pm.cfg.mu.RUnlock()
	return NvimConfigInfo{
		Bin:              pm.cfg.Nvim.Bin,
		ConfigMode:       string(pm.cfg.Nvim.ConfigMode),
		ConfigDir:        pm.cfg.Nvim.ConfigDir,
		BundledConfigDir: filepath.Join(pm.appDir, "nvim-config"),
	}
}

// ReloadConfig reloads the nvim section from a fresh config.
func (pm *PtyManager) ReloadConfig(fresh *MineoCfg) {
	pm.cfg.mu.Lock()
	defer pm.cfg.mu.Unlock()
	if fresh != nil {
		pm.cfg.Nvim.Bin = fresh.Nvim.Bin
		pm.cfg.Nvim.ConfigMode = fresh.Nvim.ConfigMode
		pm.cfg.Nvim.ConfigDir = fresh.Nvim.ConfigDir
	}
}

// GetSocketPath returns the socket path for an instance.
func (pm *PtyManager) GetSocketPath(id string) string {
	pm.mu.RLock()
	defer pm.mu.RUnlock()
	inst, ok := pm.instances[id]
	if !ok {
		return ""
	}
	return inst.socketPath
}

// DisposeAll kills all PTY instances.
func (pm *PtyManager) DisposeAll() {
	pm.mu.RLock()
	ids := make([]string, 0, len(pm.instances))
	for id := range pm.instances {
		ids = append(ids, id)
	}
	pm.mu.RUnlock()

	for _, id := range ids {
		pm.Kill(id)
	}
}

// cleanup removes a PTY from the map and handles primary promotion.
func (pm *PtyManager) cleanup(id string) {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	inst, ok := pm.instances[id]
	if !ok {
		return
	}

	inst.disposed = true
	inst.listenerMu.Lock()
	// Clear all listeners
	for k := range inst.listeners {
		delete(inst.listeners, k)
	}
	inst.listenerMu.Unlock()

	// Clean up socket file for editor PTYs
	if inst.socketPath != "" {
		os.Remove(inst.socketPath)
	}

	// Close ptmx (ignore errors — may already be closed)
	inst.ptmx.Close()

	// Wait for process to finish (non-blocking best effort)
	go func() {
		if inst.cmd.Process != nil {
			inst.cmd.Wait()
		}
	}()

	delete(pm.instances, id)

	// If primary was killed, promote next editor
	if pm.primaryID == id {
		pm.primaryID = ""
		for nextID, nextInst := range pm.instances {
			if nextInst.role == RoleNeovim && !nextInst.disposed {
				pm.primaryID = nextID
				break
			}
		}
	}
}
